src/audio/src/src.c

```diff
@@ -54,36 +55,44 @@
 #define MAX_FIR_DELAY_SIZE_XNCH (PLATFORM_MAX_CHANNELS * MAX_FIR_DELAY_SIZE)
 #define MAX_OUT_DELAY_SIZE_XNCH (PLATFORM_MAX_CHANNELS * MAX_OUT_DELAY_SIZE)
 
-static const struct comp_driver comp_src;
-
 LOG_MODULE_REGISTER(src, CONFIG_SOF_LOG_LEVEL);
 
-#if CONFIG_IPC_MAJOR_4
-/* e61bb28d-149a-4c1f-b709-46823ef5f5a3 */
-DECLARE_SOF_RT_UUID("src", src_uuid, 0xe61bb28d, 0x149a, 0x4c1f,
-		    0xb7, 0x09, 0x46, 0x82, 0x3e, 0xf5, 0xf5, 0xae);
-#elif CONFIG_IPC_MAJOR_3
+#if CONFIG_IPC_MAJOR_3
+static const struct comp_driver comp_src;
+
 /* c1c5326d-8390-46b4-aa47-95c3beca6550 */
 DECLARE_SOF_RT_UUID("src", src_uuid, 0xc1c5326d, 0x8390, 0x46b4,
 		    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50);
 #else
-#error "No or invalid IPC MAJOR version selected."
-#endif /* CONFIG_IPC_MAJOR_4 */
-
-DECLARE_TR_CTX(src_tr, SOF_UUID(src_uuid), LOG_LEVEL_INFO);
-
 /* src component private data */
 struct ipc4_config_src {
 	struct ipc4_base_module_cfg base;
 	uint32_t sink_rate;
 };
 
+/* e61bb28d-149a-4c1f-b709-46823ef5f5a3 */
+DECLARE_SOF_RT_UUID("src", src_uuid, 0xe61bb28d, 0x149a, 0x4c1f,
+		    0xb7, 0x09, 0x46, 0x82, 0x3e, 0xf5, 0xf5, 0xae);
+#endif
+
+DECLARE_TR_CTX(src_tr, SOF_UUID(src_uuid), LOG_LEVEL_INFO);
+
 struct comp_data {
-#if CONFIG_IPC_MAJOR_4
-	struct ipc4_config_src ipc_config;
-#else
+#if CONFIG_IPC_MAJOR_3
 	struct ipc_config_src ipc_config;
-#endif /* CONFIG_IPC_MAJOR_4 */
+	void (*src_func)(struct comp_dev *dev,
+			 const struct audio_stream __sparse_cache *source,
+			 struct audio_stream __sparse_cache *sink,
+			 int *consumed,
+			 int *produced);
+#else
+	struct ipc4_config_src ipc_config;
+	void (*src_func)(struct processing_module *mod,
+			 struct input_stream_buffer *bsource,
+			 struct output_stream_buffer *bsink,
+			 int *n_read,
+			 int *n_written);
+#endif
```

Comment:
These look very similar, are you sure we cant align them  - what did volume do ?

Reply:
@gongjun-song will also need sparse support - @lyakh fyi.

Reply:
thanks, updated